lighthouse-core/gather/gatherers/seo/font-size.js

```diff
@@ -0,0 +1,206 @@
+  return nodeInBody(node.parentNode);
+}
+
+/**
+ * Get list of all nodes from the document body.
+ *
+ * @param {!Object} driver
+ * @returns {!Array<!Node>}
+ */
+function getAllNodesFromBody(driver) {
+  return driver.getNodesInDocument()
+    .then(nodes => {
+      const nodeMap = new Map();
+      nodes.forEach(node => nodeMap.set(node.nodeId, node));
+      nodes.forEach(node => node.parentNode = nodeMap.get(node.parentId));
+      return nodes.filter(nodeInBody);
+    });
+}
+
+/**
+ * Returns effective CSS rule for given CSS property
+ *
+ * @param {!string} property CSS property name
+ * @param {!Node} node
+ * @param {!Object} matched CSS rules
+ * @returns {WebInspector.CSSStyleDeclaration}
+ */
+function getEffectiveRule(property, node, {
+  inlineStyle,
+  attributesStyle,
+  matchedCSSRules,
+  inherited,
+}) {
+  const cssModel = {
+    styleSheetHeaderForId: id => ({id}),
+  };
+
+  const nodeType = node.nodeType;
+  node.nodeType = () => nodeType;
+  const matchedStyles = new CSSMatchedStyles(
+    cssModel,
+    node,
+    inlineStyle,
+    attributesStyle,
+    matchedCSSRules,
+    null,
+    inherited,
+    null
+  );
+
+  const nodeStyles = matchedStyles.nodeStyles();
+  const matchingRule = nodeStyles.find(style => {
+    const foundProperty = style.allProperties.find(item => item.name === property);
+    // the applicable property will be the only one that isn't in the "overloaded" state.
+    return foundProperty &&
+      matchedStyles.propertyState(foundProperty) !== CSSMatchedStyles.PropertyState.Overloaded;
+  });
+
+  return matchingRule;
+}
+
+/**
+ * @param {!Node} node
+ * @returns {!number}
+ */
+function getNodeTextLength(node) {
+  return !node.nodeValue ? 0 : node.nodeValue.trim().length;
+}
+
+/**
+ * @param {!Object} driver
+ * @param {!Node} node text node
+ * @returns {WebInspector.CSSStyleDeclaration}
+ */
+function getFontSizeSourceRule(driver, node) {
+  return driver.sendCommand('CSS.getMatchedStylesForNode', {nodeId: node.nodeId})
+    .then(matchedRules => getEffectiveRule(FONT_SIZE_PROPERTY_NAME, node, matchedRules));
+}
+
+/**
+ * @param {!Object} driver
+ * @param {!Node} node text node
+ * @returns {!{fontSize: number, textLength: number, node: Node}}
+ */
+function getFontSizeInformation(driver, node) {
+  return driver.sendCommand('CSS.getComputedStyleForNode', {nodeId: node.parentId})
+    .then(result => {
+      const {computedStyle} = result;
+      const fontSizeProperty = computedStyle.find(({name}) => name === FONT_SIZE_PROPERTY_NAME);
+
+      return {
+        fontSize: parseInt(fontSizeProperty.value, 10),
+        textLength: getNodeTextLength(node),
+        node: node.parentNode,
+      };
+    });
+}
+
+/**
+ * @param {Node} node
+ * @returns {boolean}
+ */
+function isNonEmptyTextNode(node) {
+  return node.nodeType === global.Node.TEXT_NODE &&
+    !TEXT_NODE_BLOCK_LIST.has(node.parentNode.nodeName) &&
+    getNodeTextLength(node) > 0;
+}
+
+class FontSize extends Gatherer {
```

Comment:
this is pretty fancy now, think you could distill a lot of your learnings that you put in the GitHub comments into a @fileoverview code comment?

Reply:
Good idea! WDYT:

-----------

@fileoverview Extracts information about illegible text from the page.

In effort to keep this audit's execution time around 1s, maximum number of the protocol calls was limited. `CSS.getMatchedStylesForNode`, responsible for providing information needed to extract effective CSS rule, can be especially slow.
This gatherer also collects stylesheet metadata by itself, instead of relying on the `styles` gatherer which is slow (because it parses the stylesheet content).

-----------

Let me know if I should add or expand on something. Feel free to just edit this comment.

BTW how come @fileoverview Github nick is not taken!? 😉

Reply:
Sounds good, maybe mention the double falloff with visited vs. analyzed since that's where a lot of the trickiness later comes into play?

> BTW how come @fileoverview Github nick is not taken!? 😉

😆 well not every file is so complicated to need one, but this is a good case I think :)